### DISCUSSION TITLE
Discriminate output based on the input

### DISCUSSION THREAD

**Discussion by JesusTheHun** [trpc/trpc#5734] [Q&A]
When a procedure accepts a union as input, it would be nice to be able to narrow the output type based on the input.

```ts
const zCookie = z.object({ chocolateChips: z.number().min(50) });
const zBrownie = z.object({ almondChips: z.number().max(10) });
const zBiscuit = z.union([zCookie, zBrownie]);

type Cookie = z.infer<typeof zCookie>;
type Brownie = z.infer<typeof zBrownie>;

const appRouter = router({
  gimmeBiscuit: authenticatedProcedure
    .input(z.enum(['cookie', 'brownie']))
    .output(zBiscuit)
    .mutation(async ({ input }) => { /* bake some biscuits */ })
});
```

A caller asking for a `cookie` will receive an object of type `Cookie | Brownie`. While it is possible to pick the right output validator based on the input by passing a function to `output`, the output type will not be tied to the input type.

I tried to think about something that is feasible with TypeScript and came up with this API : 

```ts
const appRouter = router({
  gimmeBiscuit: authenticatedProcedure
    .inputOutput(z.literal('cookie'), zCookie)
    .inputOutput(z.literal('brownie'), zBrownie)
    .mutation(async ({ input }) => { /* bake some biscuits */ })
});
```

Another option would be to update the signature of the `output` method to allow a function parameter that bind the input type to the return type of the function through a generic, but I'm not sure of the implications of such update.

**Comment by KATT** [comment]
See 

- #2150
- #3311
